delta_Go/game_mechanics.py

```diff
@@ -0,0 +1,260 @@
+        if self.done:
+            self.player = random.choice(["black_0", "white_0"])
+        super().reset()
+
+        if self.verbose:
+            print(
+                f"Resetting Game.\nYou are playing with the {self.player[:-2]} tiles.\nBlack plays first\n\n"
+            )
+
+        if self.turn != self.player:
+            self._step(
+                self.opponent_choose_move(
+                    observation=self.observation, legal_moves=self.legal_moves, env=self
+                ),
+            )
+
+        return self.observation, 0, self.done, self.info
+
+    def move_to_string(self, move: int):
+        N = self.observation.shape[0]
+        if move == N**2:
+            return "passes"
+        return f"places counter at coordinate: {(move//N, move%N)}"
+
+    def __str__(self):
+        return str(self.observation) + "\n"
+
+    def _step(self, move: int) -> float:
+
+        assert not self.done, "Game is done! Please reset() the env before calling step() again"
+        assert move in self.legal_moves, f"{move} is an illegal move"
+
+        if self.verbose:
+            print(f"{self.turn_pretty} {self.move_to_string(move)}")
+
+        self.env.step(move)
+
+        if self.render:
+            self.render_game()
+
+        return self.reward
+
+    @property
+    def reward(self):
+        return self.env.last()[1]
+
+    @property
+    def player_score(self):
+        black_score = self.env.env.env.env.go_game.score()  # lol
+        return black_score if self.player == "black_0" else black_score * -1
+
+    def step(self, move: int):
+
+        # Flipped because the env takes the step, changes the player, then we return the reward
+        reward = -self._step(move)
+
+        if not self.done:
+            opponent_reward = self._step(
+                self.opponent_choose_move(
+                    observation=self.observation, legal_moves=self.legal_moves, env=self
+                ),
+            )
+            # Flipped as above
+            reward = opponent_reward
+
+        if self.verbose and self.done:
+            white_idx = int(self.turn_pretty == "white")
+            black_idx = int(self.turn_pretty == "black")
+            white_count = np.sum(self.env.last()[0]["observation"].astype("int")[:, :, white_idx])
+            black_count = np.sum(self.env.last()[0]["observation"].astype("int")[:, :, black_idx])
+
+            print(
+                f"\nGame over. Reward = {reward}.\n"
+                f"Player was playing as {self.player[:-2]}.\n"
+                f"White has {white_count} counters.\n"
+                f"Black has {black_count} counters.\n"
+                f"Your score is {self.player_score}.\n"
+            )
+
+        return self.observation, reward, self.done, self.info
+
+
+def GoEnv(
+    opponent_choose_move: Callable[[np.ndarray, np.ndarray], int],
+    verbose: bool = False,
+    render: bool = False,
+    game_speed_multiplier: int = 1,
+) -> DeltaEnv:
```

Comment:
Personally, I found this a little confusing when writing my bot. The function name looks like a class, but it isn't.

I'd either go fully functional and therefore lowercase (e.g. `get_go_env()`) or do the class thing properly:

```
class GoEnv(DeltaEnv):
    def __init__(...):
        super().__init__(...)
```